docs/docs/specs/v0/composer.md

```diff
@@ -0,0 +1,91 @@
+# Composer
+
+A developer-first, peer-to-peer knowledge management system.
+
+## Scenarios
+
+1. Users can collaborate on a github issue together
+2. Users can work on folders of markdown files
+3. Developers can extend Composer with custom data sources and visual surfaces
+
+## Extensibility
+> "Composer Chrome" refers to the total composer UI organization and all affordances surrounding the content the users interact with - all UI which is not the content itself.
+
+Two levels of complexity are proposed in sequence:
+
+**Static chrome**, where we constrain the application chrome to be statically defined and not extensible by plugins. Elements of the chrome may each expose some API surface for plugins to extend, but the overall elements of the chrome are fixed and can be assumed by all plugins without explicit dependency expressions.
+
+**Dynamic chrome**, where the chrome and it's organization is also extensible and now plugins have to express explicitly the elements of chrome they depend on because their presence can no longer be assumed.
+
+These are in addition to another orthogonal dimension of complexity, the way we load plugins:
+
+**Static plugins**, where a `<Composer />` element is given a list of `Plugin[]` directly as props and has to be compiled into the app at build time.
+
+**Dynamic plugins**, where the app knows how to load plugins via ESM `import` calls in the browser at runtime.
+
+
+### 1. Static chrome (v1)
+
+Large screens are split vertically into `sidebar` and `content` areas.
+```
+.------------.
+|    |     o |
+|    |       |
+|    |       |
+|    |       |
+'------------'
+```
+Small screens turn the `sidebar` into a slide-over element that hides by default.
+
+The HALO button (o) is in one of the corners and can be used to access the DXOS Shell.
+
+The sidebar presents a tree of nodes which can be populated by plugins.
+
+The content area can be filled by plugins, and can sense the state of the tree view (selection, nodes, etc).
+
+```ts
+
+type MaybePromise<T> = T | Promise<T>;
+
+interface Plugin {
```

Comment:
One surface that I think may be missing here is how are new nodes created? Composer needs to be able to create new objects in the space and plugins probably need to be able to teach it how to create new types of objects

Reply:
through actions (which are not yet defined in the spec) - presumably actions might submit new ECHO objects into the space, and presumably the things returned from `getTreeNodes` are somehow observable, so they would fire with new nodes when their internal queries get refreshed as a result of adding ECHO objects.

Reply:
```typescript
type Plugin = {
  manifest: { // Manifest is serializable and could be stored in the space that makes use of the plugin for example.
      id: string
      name: string
      description: string
  },
  runtime: { // or name it`extension` ?
     sidebar: {  // Separating extensibility points into logical groups. Plugins can implement a subset of them.
           // plugins can provide TreeNodes to the Tree
          getTreeNodes(parent: TreeNode): MaybePromise<TreeNode[]>;
          // plugins can provide actions to TreeNodes
          getTreeNodeActions(parent: TreeNode): MaybePromise<Action[]>;
     },
     content: {
         // plugins can provide a UI for the content area
         // content area components access the rest of the UI state via context
         getComponent(selection: TreeNode[]): MaybePromise<React.FC>;
     }
  }
}
```

Reply:
For future we might split `manifest` and `runtime` into separate JS bundles so that manifests could quickly be loaded by web apps without requiring loading code for compoenents

Reply:
recommend `meta` instead of `manifest`

Reply:
recommend `extensions` instead of `runtime`

Reply:
> recommend `extensions` instead of `runtime`

"modules"? I don't mind "extensions" either, it just feels a little weird to me to nest "extension" and "plugin" because they are synonyms

Reply:
vscode uses something like `provides` here, but fair point